mutt_thread.c

```diff
@@ -51,9 +51,130 @@ struct ThreadsContext
+  // clang-format off
+  { "unset",         UT_UNSET },   /* limited use; see validator */
+  { "flat",          UT_FLAT },
+  { "threads",       UT_THREADS },
+  { "reverse",       UT_REVERSE },
+  // aliases
+  { "no",            UT_FLAT },
+  { "yes",           UT_THREADS },
+  { NULL,            0 },
+  // clang-format on
+};
+
+struct EnumDef UseThreadsTypeDef = {
+  "use_threads_type",
+  4,
+  (struct Mapping *) &UseThreadsMethods,
+};
+
+/**
+ * mutt_thread_style - Which threading style is active?
+ * @retval UT_FLAT    No threading in use (includes use_threads=UT_UNSET)
+ * @retval UT_THREADS Normal threads (root above subthread)
+ * @retval UT_REVERSE Reverse threads (subthread above root)
+ */
+enum UseThreads mutt_thread_style(void)
+{
+  const unsigned char c_use_threads =
+      cs_subset_enum(NeoMutt->sub, "use_threads");
+  const short c_sort = cs_subset_sort(NeoMutt->sub, "sort");
+  if (c_use_threads > UT_FLAT)
+    return c_use_threads;
+  if ((c_sort & SORT_MASK) != SORT_THREADS)
+    return UT_FLAT;
+  if (c_sort & SORT_REVERSE)
+    return UT_REVERSE;
+  return UT_THREADS;
+}
+
+/**
+ * get_use_threads_str - Convert UseThreads enum to string
+ * @param value Value to convert
+ * @return string form of value
+ */
+const char *get_use_threads_str(enum UseThreads value)
+{
+  return mutt_map_get_name(value, UseThreadsMethods);
+}
+
+
+/**
+ * check_sort_use_threads - Validate combinations of "use_threads" and "sort"
+ * @param threads proposed new value of $use_threads
+ * @param sort    proposed new value of $sort
+ * @param change  name of config variable with proposed change
+ * @param other   name of other config variable that might prevent change
+ * @param err     buffer for reporting errors
+ * @return CSR_SUCCESS     change was acceptable
+ * @return CSR_ERR_INVALID change was rejected and @a err populated
+ */
+bool check_sort_use_threads(unsigned char threads, short sort, const char *change,
+                            const char *other, struct Buffer *err)
+{
+  if ((sort & SORT_MASK) != SORT_THREADS)
+    return CSR_SUCCESS;
+
+  if ((threads > UT_FLAT) && (!(sort & SORT_REVERSE) == (threads == SORT_THREADS)))
+  {
+    mutt_buffer_printf(err, _("Changing option %s would conflict with current setting of %s"),
+                       change, other);
+    return CSR_ERR_INVALID;
+  }
+
+  if (threads != UT_UNSET)
+  {
+    static bool warned = false;
+    if (!warned)
+    {
+      mutt_warning(_("Enabling threads should prefer $use_threads over $sort"));
+      warned = true;
+    }
+  }
+  return CSR_SUCCESS;
+}
+
+/**
+ * use_threads_validator - Validate values of "use_threads" - Implements ConfigDef::validator()
+ */
+int use_threads_validator(const struct ConfigSet *cs, const struct ConfigDef *cdef,
+                          intptr_t value, struct Buffer *err)
+{
+  if (value == UT_UNSET)
+  {
+    mutt_buffer_printf(err, _("Must use 'reset all' to unset %s once it is set"), cdef->name);
```

Comment:
Hmm... the **nuclear** option!
I'm not sure we should recommend that to the users.

I've been having some thoughts about 'reset'...

For the config, does 'reset' ever need a validator?
The config variable's value would either be:
1. a default, written in the code (safe)
2. a value set by the user, which passed validation (safe)

For inherited config (as yet unused), 'reset' just exposes a higher-level value which is subject to 1) and 2)

hmm... assuming that the default value is **always** valid

Reply:
You have a point. Even in the case of $sort and $use_threads having disallowed combinations, we at least know that testing just one in isolation still works (because it matches the user having just set the other, and the forbidden combo requires the user to try to set both at once). So while `set use_threads=unset` is undesirable, allowing `reset use_threads` could be okay

Reply:
as long as 'reset all' temporarily disables the validators, the check for UT_UNSET can only be reached during 'set use_threads=unset' and 'reset use_threads'.  But the validator can't tell the difference, without a change to the signature.  Thus, documenting 'reset all' is correct, but you are right that it is rather powerful.  I'm still thinking about it